jax/_src/tree_util.py

```diff
@@ -267,6 +267,22 @@ def tree_all(tree):
   lambda s, values: collections.defaultdict(s[0], safe_zip(s[1], values)))  # type: ignore[index]
 
 
+
+class _HashableCallableShim:
+  """Object that delegates __call__, __hash__, and __eq__ to another object."""
+  def __init__(self, fun):
+    self.fun = fun
+
+  def __call__(self, *args, **kw):
+    return self.fun(*args, **kw)
+
+  def __hash__(self):
+    return hash(self.fun)
+
+  def __eq__(self, other):
+    return self.fun == other
```

Comment:
```suggestion
  def __eq__(self, other):
    if isinstance(other, _HashableCallableShim):
      other = other.fun
    return self.fun == other
```

maybe also add this?